### PR TITLE
Small experiment concerning hydration feedback in dev mode (issue #598)

### PR DESCRIPTION
**Pull request by matsdahlin**
> ## Note: This PR is _not_ meant to be merged, it's just to share idea and keep the conversation going

## Changes

When running Astro in development mode, adds a small "devtool" inspired by Nextjs dev badge. This badge can toggle an "x-ray"-view of sorts to see which components are static and not, and which hydration mode is being used.

Also adds a click handler when devs try to interact with static components, informing them that they need to hydrate.

astro-hydration-feedback

## Testing

Copied one of the example projects under /examples and added a couple of the <Counter /> components that come with Astro templates.

## Docs

This is just hacking / throwing around ideas

### CODE REVIEW

**Comment by jasikpark** [comment]
That looks pretty great! How does this code interact with a form that's meant to submit to a path w/o javascript though?

**Comment by MatsDahlin** [comment]
> That looks pretty great! How does this code interact with a form that's meant to submit to a path w/o javascript though?

Thanks!

I pushed a small update with a suggestion of how to handle this scenario. 
If the element in the static form has a <form> ancestor, the feedback won't be shown. It will still show if you click outside a <form> element, but still within the static component. 

This feedback makes sense to me, what do you think?

astro-hydration-feedback-with-form

**Comment by jasikpark** [comment]
> > That looks pretty great! How does this code interact with a form that's meant to submit to a path w/o javascript though?
> 
> Thanks!
> 
> I pushed a small update with a suggestion of how to handle this scenario.
> If the element in the static form has a `<form>` ancestor, the feedback won't be shown. It will still show if you click outside a `<form>` element, but still within the static component.
> 
> This feedback makes sense to me, what do you think?
> 
> Here's a demo gif, i've tested hooking up a submit button aswell and it maintains default browser behaviour as far as I can see.
> 
> astro-hydration-feedback-with-form

That looks great to me :)

**Comment by jasikpark** [comment]
I think it'd be cool to be able to enable/disable this via cli flag?
```shell
$ yarn dev --debug
and 
$ yarn dev --no-debug
```
perhaps?

**Comment by stramel** [comment]
> I think it'd be cool to be able to enable/disable this via cli flag?
> 
> ```shell
> $ yarn dev --debug
> and 
> $ yarn dev --no-debug
> ```
> 
> perhaps?

+1 to opting in to enable/disable this feature.

I like the idea of `yarn dev --debug` or providing an in-browser global function to enable it something like `Astro.debug()`.

**Comment by stramel** [comment]
I also like the idea of it suggesting devtools extensions (I believe @jasikpark suggested this?). Might be able to do so based upon the components being used on the page and outputting it into the devtools console or in the terminal at boot?

**Comment by jasikpark** [comment]
> I also like the idea of it suggesting devtools extensions (I believe @jasikpark suggested this?). Might be able to do so based upon the components being used on the page and outputting it into the devtools console or in the terminal at boot?

Well 😅 I was suggesting putting a reference to the devtools extensions A. in the section of the docs for each renderer, B. Underneath the docs that would mention this feature, though I could see suggesting devtools based on which renderers you are using might be useful.

**Comment by MatsDahlin** [comment]
I agree about the CLI flag(s) and finally had some time to play around with this. 

I updated the branch with a [commit](https://github.com/snowpackjs/astro/pull/801/commits/f7076e189ad65904f9c6dd7e53743f4da7ff5be8) where I added a `--debug` flag to the CLI, and now that flag controls whether the "devtools" are injected or not, like @jasikpark suggested.

For now I only added `--debug` and _not_ `--no-debug`, because i'm not sure the latter is needed. If someone has a scenario where the `--no-debug` would be useful, please share! It would be easy to add that too.

File: packages/astro/src/frontend/dev-tools.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  allUnhydratedComponents.forEach((componentElement) => {
+    const componentEl = componentElement as HTMLElement;
+    const componentName = componentEl.dataset.astroComponentName;
+    componentEl.addEventListener('click', (e) => {
+      const elementHasFormParent = e.composedPath().filter((pathNode) => (pathNode as HTMLElement).tagName === 'FORM').length > 0;
+
+      if (!elementHasFormParent) {
+        componentElement.classList.add('show-notification');
+        console.warn(
+          `component < ${componentName} /> is being statically rendered. Use the client:* directives to hydrate it. Documentation: https://astro-docs-e7osusbql-pikapkg.vercel.app/core-concepts/component-hydration`
```
Comment (jasikpark) [inline packages/astro/src/frontend/dev-tools.ts:113]: probably want to update the url for the docs here to https://docs.astro.build, right?

Comment (MatsDahlin) [reply packages/astro/src/frontend/dev-tools.ts]: Good catch! Thanks for the input. I pushed a change for this.

**Comment by matthewp** [comment]
I think this is really cool! As the PR is not intended to be merged I'm going to close it, however. I think the next step here is to create an RFC with the proposed changes (if it's just this exact feature that's ok), and then from there, assuming it gets accepted, the existing code can be used to create a mergeable PR.

**Pull request closed without merging**